src/meta/src/manager/catalog/fragment.rs

```diff
@@ -262,6 +263,94 @@ where
+        // workaround for not having the version info in the fragment manager.
+        let old_table_fragment = table_fragments
+            .remove(table_id)
+            .with_context(|| format!("table_fragment not exist: id={}", table_id))?;
+        let mut table_fragment = table_fragments
+            .remove(dummy_table_id)
+            .with_context(|| format!("table_fragment not exist: id={}", dummy_table_id))?;
+
+        assert_eq!(table_fragment.state(), State::Initial);
+        table_fragment.set_table_id(table_id);
+
+        // Directly set to `Created` and `Running` state.
+        table_fragment.set_state(State::Created);
+        table_fragment.update_actors_state(ActorState::Running);
+
+        table_fragments.insert(table_id, table_fragment.clone());
+
+        // Update downstream `Merge`s.
+        let mut merge_updates: HashMap<_, _> = merge_updates
+            .iter()
+            .map(|update| (update.actor_id, update))
+            .collect();
+
+        let to_update_table_ids = table_fragments
+            .tree_ref()
+            .iter()
+            .filter(|(_, v)| {
+                v.actor_ids()
+                    .iter()
+                    .any(|&actor_id| merge_updates.contains_key(&actor_id))
+            })
+            .map(|(k, _)| *k)
+            .collect::<Vec<_>>();
+
+        for table_id in to_update_table_ids {
+            let mut table_fragment = table_fragments
+                .get_mut(table_id)
+                .with_context(|| format!("table_fragment not exist: id={}", table_id))?;
+
+            for actor in table_fragment
+                .fragments
+                .values_mut()
+                .flat_map(|f| &mut f.actors)
+            {
+                if let Some(merge_update) = merge_updates.remove(&actor.actor_id) {
+                    assert!(merge_update.removed_upstream_actor_id.is_empty());
+                    assert!(merge_update.new_upstream_fragment_id.is_some());
+
+                    let stream_node = actor.nodes.as_mut().unwrap();
+                    visit_stream_node(stream_node, |body| {
+                        if let NodeBody::Merge(m) = body
+                           && m.upstream_fragment_id == merge_update.upstream_fragment_id
+                        {
+                            m.upstream_fragment_id = merge_update.new_upstream_fragment_id.unwrap();
+                            m.upstream_actor_id = merge_update.added_upstream_actor_id.clone();
+                        }
+                    });
+                }
+            }
+        }
+
+        assert!(merge_updates.is_empty());
+
+        // Commit changes and notify about the changes.
+        commit_meta!(self, table_fragments)?;
+
+        self.notify_fragment_mapping(&old_table_fragment, Operation::Delete)
+            .await;
+        self.notify_fragment_mapping(&table_fragment, Operation::Add)
+            .await;
```

Comment:
Can this 2 lines be merged into 1 Update?

Reply:
I guess we can't. By specifying `Update` we will update the vnode mapping for all fragment in the list that is used for scaling, while in this case, the fragment list is replaced and we should delete the entry of vnode mapping for old fragments.